### PR TITLE
eslint: Enable more react rules

### PR DESCRIPTION
**Pull request by anomiex**
## Proposed changes:

* `react-hooks/exhaustive-deps`: Apply to `useSelect`.
* `react/no-direct-mutation-state`
* `react/no-find-dom-node`
* `react/no-unescaped-entities`

The first of those accounts for the majority of fixes in this PR, we had a bunch of places either missing a dependency or specifying a dependency unnecessarily. Of particular note is that constants defined outside of the `render()` function, like `STORE_ID`, don't need to be dependencies.

Of the rest, `react/no-direct-mutation-state` was only violated in one place, `react/no-unescaped-entities` in one storybook story, and `react/no-find-dom-node` was already fixed everywhere.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

Followup to https://github.com/Automattic/jetpack/pull/38436

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* CI happy?
* Do the components modified here behave as they did before, or better?
  * The `useSelect` dependency removals shouldn't have any effect; if they do there's probably something more wrong with the underlying logic and a bug with delayed updates anyway.
  * The additions to `useSelect`, if they do anything, should actually fix bugs with things sometimes not being updated. But more likely the added stuff is effectively constant.
  * Same goes for the `setState` calls, if anything it should fix bugs where changes got overwritten by later state updates.

### CODE REVIEW

File: tools/js-tools/eslintrc/react.js [hunk 1]
```diff
@@ -23,10 +23,6 @@ module.exports = {
 		'react/prefer-es6-class': 'warn',
 
 		// Temporarily override plugin:@wordpress/react so we can clean up failing stuff in separate PRs.
-		'react-hooks/exhaustive-deps': [ 'warn', { additionalHooks: '' } ],
 		'react/jsx-key': 'off',
-		'react/no-direct-mutation-state': 'off',
-		'react/no-find-dom-node': 'off',
```
Comment (tbradsha) [inline tools/js-tools/eslintrc/react.js:29]: Did I miss something or was this never causing issues to begin with?

Comment (tbradsha) [reply tools/js-tools/eslintrc/react.js:29]: Ah, you mention that in the PR body. Helps to read the PR before jumping into the code, I guess...

**Review by tbradsha — APPROVED** [review]
My React is rusty, but looks good to me.

I'll note that in general I don't like the changes we made here to appease `react/no-unescaped-entities` (it makes it harder to search for the string), but in this case it only affected Storybook, so meh.

File: tools/js-tools/eslintrc/react.js (hunk 1 above)
Comment (anomiex) [reply tools/js-tools/eslintrc/react.js:29]: Back in #38436 there was one thing left triggering it, but that got fixed by #38339.

**Comment by anomiex** [comment]
Outside of stories any string that might trigger this is going to have to be wrapped in `__()` anyway so it shouldn't come up. That's most of the reason I went for it.

**Pull request merged**